client/consensus/grandpa/authorities.go

```diff
@@ -0,0 +1,814 @@
+	sas.mtx.Lock()
+	defer sas.mtx.Unlock()
+	return sas.inner.applyStandardChanges(finalizedHash, finalizedNumber, isDescendentOf, telemetry)
+}
+
+// EnactsStandardChange Check whether the given finalized block number enacts any standard
+// authority set change (without triggering it), ensuring that if there are
+// multiple changes in the same branch, finalizing this block won't
+// finalize past multiple transitions (i.e. transitions must be finalized
+// in-order). Returns *true if the block being finalized enacts a
+// change that can be immediately applied, *false if the block being
+// finalized enacts a change but it cannot be applied yet since there are
+// other dependent changes, and nil if no change is enacted. The given
+// function `is_descendent_of` should return `true` if the second hash
+// (target) is a descendent of the first hash (base).
+func (sas *SharedAuthoritySet[H, N]) EnactsStandardChange(finalizedHash H, finalizedNumber N, isDescendentOf IsDescendentOf[H]) (*bool, error) {
+	sas.mtx.Lock()
+	defer sas.mtx.Unlock()
+	return sas.inner.EnactsStandardChange(finalizedHash, finalizedNumber, isDescendentOf)
+}
+
+// Status of the set after changes were applied.
+type Status[H comparable, N constraints.Unsigned] struct {
+	// Whether internal changes were made.
+	changed bool
+	// Not nil when underlying authority set has changed, containing the
+	// block where that set changed.
+	newSetBlock *hashNumber[H, N]
+}
+
+// AuthoritySet A set of authorities.
+type AuthoritySet[H comparable, N constraints.Unsigned] struct {
+	// The current active authorities.
+	currentAuthorities AuthorityList
+	// The current set id.
+	setId uint64
+	// Tree of pending standard changes across forks. Standard changes are
+	// enacted on finality and must be enacted (i.e. finalized) in-order across
+	// a given branch
+	pendingStandardChanges ForkTree[H, N]
+	// Pending forced changes across different forks (at most one per fork).
+	// Forced changes are enacted on block depth (not finality), for this
+	// reason only one forced hashNumber should exist per fork. When trying to
+	// apply forced changes we keep track of any pending standard changes that
+	// they may depend on, this is done by making sure that any pending hashNumber
+	// that is an ancestor of the forced changed and its effective block number
+	// is lower than the last finalized block (as signaled in the forced
+	// hashNumber) must be applied beforehand.
+	pendingForcedChanges []PendingChange[H, N]
+	// Track at which blocks the set id changed. This is useful when we need to prove finality for
+	// a given block since we can figure out what set the block belongs to and when the set
+	// started/ended.
+	authoritySetChanges AuthoritySetChanges[N]
+}
+
+// InvalidAuthorityList authority sets must be non-empty and all weights must be greater than 0
+func InvalidAuthorityList(authorities AuthorityList) bool {
```

Comment:
don't think we need to make this public.  
```suggestion
func invalidAuthorityList(authorities AuthorityList) bool {
```